{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "38ca2770",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Features\n",
    "# GC Content\n",
    "# CpG Islands\n",
    "# CpG Islands per ORF Length\n",
    "# Transcript Length\n",
    "# Accumulated Nucleotide Frequency A\n",
    "# Accumulated Nucleotide Frequency G\n",
    "# Accumulated Nucleotide Frequency T\n",
    "# Accumulated Nucleotide Frequency C\n",
    "# Relative Codon Bias\n",
    "# Stop Codon Frequency\n",
    "# Isoelectric Point\n",
    "# Molecular Weight\n",
    "# Gravy\n",
    "# Instability Index\n",
    "# Stop Frame Score\n",
    "# Stop Codon Frequency\n",
    "# GC per ORF Length\n",
    "# ORF Length\n",
    "# Fickett Score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08e328d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.SeqUtils import ProtParam\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f805ca78",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_file(filename):\n",
    "    seq_tuple = []\n",
    "    count = 0\n",
    "    with open(filename) as f:\n",
    "        header = \"\"\n",
    "        sequence = \"\"\n",
    "        for line in f:\n",
    "            line = line.strip()  #remove white spaces\n",
    "            \n",
    "            if line.startswith('>'):\n",
    "                if sequence != '':\n",
    "                    #to filter sequences by their length\n",
    "#                     seq_len = len(sequence)\n",
    "#                     if seq_len >= 200 and seq_len <= 3000:\n",
    "                        seq_tuple.append((header, sequence))\n",
    "                header = line[1:]\n",
    "                sequence = ''\n",
    "                count += 1\n",
    "                \n",
    "            else:\n",
    "                sequence += line\n",
    "                \n",
    "        # Check the last sequence\n",
    "        seq_len = len(sequence)\n",
    "        if seq_len >= 300 and seq_len <= 1000:\n",
    "            seq_tuple.append((header, sequence))\n",
    "        return seq_tuple, count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8c2ab12c",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding = \"coding.fa\"\n",
    "non_coding = \"noncoding.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4ec550a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_list, coding_count = load_file(coding)  # function call on coding dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ccee9f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    " noncoding_list, noncoding_count = load_file(non_coding)   # function call on noncoding dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ffbef914",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "173111"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(noncoding_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9564c6c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "111275"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(coding_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "378e4512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "111276"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "702dc182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "173112"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "noncoding_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "546c308b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba83045c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aad4a350",
   "metadata": {},
   "source": [
    "## Fickett Score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b6017518",
   "metadata": {},
   "outputs": [],
   "source": [
    "def look_up_position_prob(value, base, position_para, position_prob, position_weight):\n",
    "\n",
    "\t\"\"\"look up positional probability by base and value\"\"\"\n",
    "\n",
    "\tif float(value) < 0:\n",
    "\t\treturn None\n",
    "\tfor idx, val in enumerate(position_para):\n",
    "\t\tif float(value) >= val:\n",
    "\t\t\treturn float(position_prob[base][idx]) * float(position_weight[base])\n",
    "\n",
    "\n",
    "def look_up_content_prob(value, base, content_para, content_prob, content_weight):\n",
    "\n",
    "\t\"\"\"look up content probability by base and value\"\"\"\n",
    "\n",
    "\tif float(value) < 0:\n",
    "\t\treturn None\n",
    "\tfor idx, val in enumerate(content_para):\n",
    "\t\tif float(value) >= val:\n",
    "\t\t\treturn float(content_prob[base][idx]) * float(content_weight[base])\n",
    "        \n",
    "def fickett_value_full_sequence(seq, type_seq):\n",
    "\n",
    "\t\"\"\"calculate Fickett from full sequence - CPC2\"\"\"\n",
    "\n",
    "\tposition_para = [1.9, 1.8, 1.7, 1.6, 1.5, 1.4, 1.3, 1.2, 1.1, 0.0]\n",
    "\tcontent_para = [0.33, 0.31, 0.29, 0.27, 0.25, 0.23, 0.21, 0.19, 0.17, 0]\n",
    "\n",
    "\tposition_prob = {\n",
    "\t\t'A': [0.51, 0.55, 0.57, 0.52, 0.48, 0.58, 0.57, 0.54, 0.50, 0.36],\n",
    "\t\t'C': [0.29, 0.44, 0.55, 0.49, 0.52, 0.60, 0.60, 0.56, 0.51, 0.38],\n",
    "\t\t'G': [0.62, 0.67, 0.74, 0.65, 0.61, 0.62, 0.52, 0.41, 0.31, 0.17],\n",
    "\t\t'T': [0.51, 0.60, 0.69, 0.64, 0.62, 0.67, 0.58, 0.48, 0.39, 0.24]}\n",
    "\t\n",
    "\tposition_weight = {'A': 0.062, 'C': 0.093, 'G': 0.205, 'T': 0.154}\n",
    "\tcontent_weight = {'A': 0.084, 'C': 0.076, 'G': 0.081, 'T': 0.055}\n",
    "\n",
    "\tcontent_prob = {\n",
    "\t\t'A': [0.40, 0.55, 0.58, 0.58, 0.52, 0.48, 0.45, 0.45, 0.38, 0.19],\n",
    "\t\t'C': [0.50, 0.63, 0.59, 0.50, 0.46, 0.45, 0.47, 0.56, 0.59, 0.33],\n",
    "\t\t'G': [0.21, 0.40, 0.47, 0.50, 0.52, 0.56, 0.57, 0.52, 0.44, 0.23],\n",
    "\t\t'T': [0.30, 0.49, 0.56, 0.53, 0.48, 0.48, 0.52, 0.57, 0.60, 0.51]}\n",
    "\n",
    "\tif len(seq) < 2:\n",
    "\t\treturn 0\n",
    "\n",
    "\tfickett_score = 0\n",
    "\tseq = seq.upper()\n",
    "\ttotal_base = len(seq)\n",
    "\n",
    "\tphase_0 = seq[::3]\n",
    "\tphase_1 = seq[1::3]\n",
    "\tphase_2 = seq[2::3]\n",
    "\n",
    "\tphase_0_A = phase_0.count('A')\n",
    "\tphase_1_A = phase_1.count('A')\n",
    "\tphase_2_A = phase_2.count('A')\n",
    "\tphase_0_C = phase_0.count('C')\n",
    "\tphase_1_C = phase_1.count('C')\n",
    "\tphase_2_C = phase_2.count('C')\n",
    "\tphase_0_G = phase_0.count('G')\n",
    "\tphase_1_G = phase_1.count('G')\n",
    "\tphase_2_G = phase_2.count('G')\n",
    "\tif type_seq == 1:\n",
    "\t\tphase_0_T = phase_0.count('T')\n",
    "\t\tphase_1_T = phase_1.count('T')\n",
    "\t\tphase_2_T = phase_2.count('T')\n",
    "\telse:\n",
    "\t\tphase_0_T = phase_0.count('U')\n",
    "\t\tphase_1_T = phase_1.count('U')\n",
    "\t\tphase_2_T = phase_2.count('U')\n",
    "\n",
    "\tA_content = float(phase_0_A + phase_1_A + phase_2_A) / total_base\n",
    "\tC_content = float(phase_0_C + phase_1_C + phase_2_C) / total_base\n",
    "\tG_content = float(phase_0_G + phase_1_G + phase_2_G) / total_base\n",
    "\tT_content = float(phase_0_T + phase_1_T + phase_2_T) / total_base\n",
    "\tA_position = max([phase_0_A, phase_1_A, phase_2_A]) / (min([phase_0_A, phase_1_A, phase_2_A]) + 1.0)\n",
    "\tC_position = max([phase_0_C, phase_1_C, phase_2_C]) / (min([phase_0_C, phase_1_C, phase_2_C]) + 1.0)\n",
    "\tG_position = max([phase_0_G, phase_1_G, phase_2_G]) / (min([phase_0_G, phase_1_G, phase_2_G]) + 1.0)\n",
    "\tT_position = max([phase_0_T, phase_1_T, phase_2_T]) / (min([phase_0_T, phase_1_T, phase_2_T]) + 1.0)\n",
    "\n",
    "\tfickett_score += look_up_content_prob(A_content, 'A', content_para, content_prob, content_weight)\n",
    "\tfickett_score += look_up_content_prob(C_content, 'C', content_para, content_prob, content_weight)\n",
    "\tfickett_score += look_up_content_prob(G_content, 'G', content_para, content_prob, content_weight)\n",
    "\tfickett_score += look_up_content_prob(T_content, 'T', content_para, content_prob, content_weight)\n",
    "\n",
    "\tfickett_score += look_up_position_prob(A_position, 'A', position_para, position_prob, position_weight)\n",
    "\tfickett_score += look_up_position_prob(C_position, 'C', position_para, position_prob, position_weight)\n",
    "\tfickett_score += look_up_position_prob(G_position, 'G', position_para, position_prob, position_weight)\n",
    "\tfickett_score += look_up_position_prob(T_position, 'T', position_para, position_prob, position_weight)\n",
    "\n",
    "\treturn fickett_score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98f56abe",
   "metadata": {},
   "source": [
    "## Accumulated Nucleotide Frequency for A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dcef6260",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def accumulated_nucleotide_frequency(sequence):\n",
    "    sequence = sequence.upper()\n",
    "    mapping = {'A': [], 'C': [], 'T': [], 'G': []}\n",
    "    counts = {'A': 0, 'C': 0, 'T': 0, 'G': 0}\n",
    "\n",
    "    for i in range(len(sequence)):\n",
    "        nucleotide = sequence[i]\n",
    "        counts[nucleotide] += 1\n",
    "        for nt, count in counts.items():\n",
    "            mapping[nt].append(count / (i + 1))\n",
    "\n",
    "    return mapping\n",
    "\n",
    "def accumulated_frequency_A(sequence):\n",
    "    return accumulated_nucleotide_frequency(sequence)['A'][-1]\n",
    "\n",
    "# def accumulated_frequency_C(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['C']\n",
    "\n",
    "# def accumulated_frequency_T(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['T']\n",
    "\n",
    "# def accumulated_frequency_G(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['G']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8778fc0d",
   "metadata": {},
   "source": [
    "## Accumulated Nucleotide Frequency for G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "697c42eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def accumulated_nucleotide_frequency(sequence):\n",
    "    sequence = sequence.upper()\n",
    "    mapping = {'A': [], 'C': [], 'T': [], 'G': []}\n",
    "    counts = {'A': 0, 'C': 0, 'T': 0, 'G': 0}\n",
    "\n",
    "    for i in range(len(sequence)):\n",
    "        nucleotide = sequence[i]\n",
    "        counts[nucleotide] += 1\n",
    "        for nt, count in counts.items():\n",
    "            mapping[nt].append(count / (i + 1))\n",
    "\n",
    "    return mapping\n",
    "\n",
    "def accumulated_frequency_G(sequence):\n",
    "    return accumulated_nucleotide_frequency(sequence)['G'][-1]\n",
    "\n",
    "# def accumulated_frequency_C(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['C']\n",
    "\n",
    "# def accumulated_frequency_T(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['T']\n",
    "\n",
    "# def accumulated_frequency_G(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['G']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c493970",
   "metadata": {},
   "source": [
    "## Accumulated Nucleotide Frequency for T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db3e6d4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def accumulated_nucleotide_frequency(sequence):\n",
    "    sequence = sequence.upper()\n",
    "    mapping = {'A': [], 'C': [], 'T': [], 'G': []}\n",
    "    counts = {'A': 0, 'C': 0, 'T': 0, 'G': 0}\n",
    "\n",
    "    for i in range(len(sequence)):\n",
    "        nucleotide = sequence[i]\n",
    "        counts[nucleotide] += 1\n",
    "        for nt, count in counts.items():\n",
    "            mapping[nt].append(count / (i + 1))\n",
    "\n",
    "    return mapping\n",
    "\n",
    "def accumulated_frequency_T(sequence):\n",
    "    return accumulated_nucleotide_frequency(sequence)['T'][-1]\n",
    "\n",
    "# def accumulated_frequency_C(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['C']\n",
    "\n",
    "# def accumulated_frequency_T(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['T']\n",
    "\n",
    "# def accumulated_frequency_G(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['G']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dcaddd2",
   "metadata": {},
   "source": [
    "## Accumulated Nucleotide Frequency for C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3451bde3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def accumulated_nucleotide_frequency(sequence):\n",
    "    sequence = sequence.upper()\n",
    "    mapping = {'A': [], 'C': [], 'T': [], 'G': []}\n",
    "    counts = {'A': 0, 'C': 0, 'T': 0, 'G': 0}\n",
    "\n",
    "    for i in range(len(sequence)):\n",
    "        nucleotide = sequence[i]\n",
    "        counts[nucleotide] += 1\n",
    "        for nt, count in counts.items():\n",
    "            mapping[nt].append(count / (i + 1))\n",
    "\n",
    "    return mapping\n",
    "\n",
    "def accumulated_frequency_C(sequence):\n",
    "    return accumulated_nucleotide_frequency(sequence)['C'][-1]\n",
    "\n",
    "# def accumulated_frequency_C(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['C']\n",
    "\n",
    "# def accumulated_frequency_T(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['T']\n",
    "\n",
    "# def accumulated_frequency_G(sequence):\n",
    "#     return accumulated_nucleotide_frequency(sequence)['G']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb11910a",
   "metadata": {},
   "source": [
    "## GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b15829ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gcContent(sequence):\n",
    "    '''\n",
    "    :param sequence: primary sequence of a transcript\n",
    "    :return: the GC content of the sequence\n",
    "    '''\n",
    "    Gcontent = 0\n",
    "    Ccontent = 0\n",
    "    for base in sequence:\n",
    "        if base == 'G':\n",
    "            Gcontent += 1\n",
    "        if base == 'C':\n",
    "            Ccontent += 1\n",
    "    GCcontent = (Gcontent+Ccontent)/len(sequence)\n",
    "    return GCcontent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d3fc58f",
   "metadata": {},
   "source": [
    "## Relative Codon Bias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e7c34cab",
   "metadata": {},
   "outputs": [],
   "source": [
    "    \n",
    "def RCB_score(seq):\n",
    "    condon_dict = {}         # a dictionary contain the counts of each condon\n",
    "    condon_list = []         # a list contain the condons in a sequence\n",
    "    Frequency_1 = {'A': 0,\n",
    "                'T': 0,\n",
    "                'G' : 0,\n",
    "                'C' : 0}\n",
    "    Frequency_2 = {'A': 0,\n",
    "                'T': 0,\n",
    "                'G' : 0,\n",
    "                'C' : 0}\n",
    "    Frequency_3 = {'A': 0,\n",
    "                'T': 0,\n",
    "                'G' : 0,\n",
    "                'C' : 0}\n",
    "\n",
    "    # First position\n",
    "    for i in range(0,len(seq),3):\n",
    "        if seq[i] == 'A':\n",
    "            Frequency_1['A'] += 1\n",
    "        elif seq[i] == 'T':\n",
    "            Frequency_1['T'] += 1\n",
    "        elif seq[i] == 'G':\n",
    "            Frequency_1['G'] += 1\n",
    "        elif seq[i] == 'C':\n",
    "            Frequency_1['C'] += 1\n",
    "\n",
    "        if len(seq[i:i+3]) != 3:\n",
    "            continue\n",
    "        elif seq[i:i+3] not in condon_dict:\n",
    "            condon_dict[seq[i:i+3]] = 1\n",
    "        else:\n",
    "            condon_dict[seq[i:i+3]] += 1\n",
    "\n",
    "        if len(seq[i:i+3]) == 3:\n",
    "            condon_list.append(seq[i:i+3])\n",
    "\n",
    "    # Second position\n",
    "    for i in range(1,len(seq),3):\n",
    "        if seq[i] == 'A':\n",
    "            Frequency_2['A'] += 1\n",
    "        elif seq[i] == 'T':\n",
    "            Frequency_2['T'] += 1\n",
    "        elif seq[i] == 'G':\n",
    "            Frequency_2['G'] += 1\n",
    "        elif seq[i] == 'C':\n",
    "            Frequency_2['C'] += 1\n",
    "\n",
    "    # Third position\n",
    "    for i in range(2,len(seq),3):\n",
    "        if seq[i] == 'A':\n",
    "            Frequency_3['A'] += 1\n",
    "        elif seq[i] == 'T':\n",
    "            Frequency_3['T'] += 1\n",
    "        elif seq[i] == 'G':\n",
    "            Frequency_3['G'] += 1\n",
    "        elif seq[i] == 'C':\n",
    "            Frequency_3['C'] += 1\n",
    "\n",
    "    total_condons = len(condon_list)\n",
    "    condon_sum = 0\n",
    "    for k in condon_list:\n",
    "        d = np.log(np.absolute((condon_dict[k]/total_condons) - (Frequency_1[k[0]]/total_condons) *\n",
    "             (Frequency_2[k[1]]/total_condons) * (Frequency_3[k[2]]/total_condons)) / ((Frequency_1[k[0]]/total_condons) *\n",
    "             (Frequency_2[k[1]]/total_condons) * (Frequency_3[k[2]]/total_condons)) + 1)\n",
    "        condon_sum = condon_sum + d\n",
    "\n",
    "    RCB = np.exp(condon_sum/len(condon_list))-1\n",
    "    return RCB\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85b44334",
   "metadata": {},
   "source": [
    "## Translation function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "62e33710",
   "metadata": {},
   "outputs": [],
   "source": [
    "def translation(sequence):\n",
    "    '''\n",
    "    :param sequence: input primary sequence of a transcript\n",
    "    :return: protein sequence\n",
    "    '''\n",
    "    trans_dic = {'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L', 'CTT': 'L',\n",
    "                 'CTC': 'L', 'CTA': 'L', 'CTG': 'L', 'ATT': 'I', 'ATC': 'I',\n",
    "                 'ATA': 'I', 'ATG': 'M', 'GTT': 'V', 'GTC': 'V', 'GTA': 'V',\n",
    "                 'GTG': 'V', 'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "                 'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P', 'ACT': 'T',\n",
    "                 'ACC': 'T', 'ACA': 'T', 'ACG': 'T', 'GCT': 'A', 'GCC': 'A',\n",
    "                 'GCA': 'A', 'GCG': 'A', 'TAT': 'Y', 'TAC': 'Y', 'TAA': '*',\n",
    "                 'TAG': '*', 'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "                 'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K', 'GAT': 'D',\n",
    "                 'GAC': 'D', 'GAA': 'E', 'GAG': 'E', 'TGT': 'C', 'TGC': 'C',\n",
    "                 'TGA': '*', 'TGG': 'W', 'CGT': 'R', 'CGC': 'R', 'CGA': 'R',\n",
    "                 'CGG': 'R', 'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "                 'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'}\n",
    "    protein = ''\n",
    "    for i in range(0, len(sequence)-2, 3):\n",
    "        protein += trans_dic[sequence[i:i+3]]\n",
    "    return protein"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dbf853d",
   "metadata": {},
   "source": [
    "## ORF Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "014c0997",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ORFfinder_T0(sequence):\n",
    "    # 3 frame translation\n",
    "    trans1 = translation(sequence)\n",
    "    trans2 = translation(sequence[1:])\n",
    "    trans3 = translation(sequence[2:])\n",
    "\n",
    "    orf1 = re.finditer(r'M.*?\\*', trans1)\n",
    "    orf2 = re.finditer(r'M.*?\\*', trans2)\n",
    "    orf3 = re.finditer(r'M.*?\\*', trans3)\n",
    "    orf1_seqs = [(m.start(), m.end(), m.group(), 'frame1') for m in orf1]\n",
    "    orf2_seqs = [(m.start(), m.end(), m.group(), 'frame2') for m in orf2]\n",
    "    orf3_seqs = [(m.start(), m.end(), m.group(), 'frame3') for m in orf3]\n",
    "\n",
    "    orfs = orf1_seqs + orf2_seqs + orf3_seqs\n",
    "    if len(orfs) == 0:\n",
    "#         return ('NaN', 'NaN'), 'NaN', 0\n",
    "        return 0\n",
    "    else:\n",
    "        orf_sorted = sorted(orfs, key=lambda t: len(t[2]), reverse=True)\n",
    "        longest_orf_protein_seq = orf_sorted[0][2]\n",
    "        longest_orf_frame = orf_sorted[0][3]\n",
    "        longest_orf_pro_position = (int(orf_sorted[0][0]), int(orf_sorted[0][1]))\n",
    "        if longest_orf_frame == 'frame1':\n",
    "            longest_orf_rna_seq_position = (longest_orf_pro_position[0] * 3, longest_orf_pro_position[1] * 3)\n",
    "            longest_orf_rna_seq = sequence[longest_orf_rna_seq_position[0]:longest_orf_rna_seq_position[1]]\n",
    "            longest_orf_rna_length = longest_orf_rna_seq_position[1] - longest_orf_rna_seq_position[0]\n",
    "            return longest_orf_rna_length\n",
    "#             return longest_orf_rna_seq_position, longest_orf_rna_seq, longest_orf_rna_length\n",
    "        elif longest_orf_frame == 'frame2':\n",
    "            longest_orf_rna_seq_position = (longest_orf_pro_position[0] * 3, longest_orf_pro_position[1] * 3)\n",
    "            longest_orf_rna_seq = sequence[1:][longest_orf_rna_seq_position[0]:longest_orf_rna_seq_position[1]]\n",
    "            longest_orf_rna_seq_original_position = (\n",
    "            longest_orf_pro_position[0] * 3 + 1, longest_orf_pro_position[1] * 3 + 1)\n",
    "            longest_orf_rna_length = longest_orf_rna_seq_original_position[1] - longest_orf_rna_seq_original_position[0]\n",
    "            return longest_orf_rna_length\n",
    "#             return longest_orf_rna_seq_original_position, longest_orf_rna_seq, longest_orf_rna_length\n",
    "        elif longest_orf_frame == 'frame3':\n",
    "            longest_orf_rna_seq_position = (longest_orf_pro_position[0] * 3, longest_orf_pro_position[1] * 3)\n",
    "            longest_orf_rna_seq = sequence[2:][longest_orf_rna_seq_position[0]:longest_orf_rna_seq_position[1]]\n",
    "            longest_orf_rna_seq_original_position = (\n",
    "            longest_orf_pro_position[0] * 3 + 2, longest_orf_pro_position[1] * 3 + 2)\n",
    "            longest_orf_rna_length = longest_orf_rna_seq_original_position[1] - longest_orf_rna_seq_original_position[0]\n",
    "            return longest_orf_rna_length\n",
    "#             return longest_orf_rna_seq_original_position, longest_orf_rna_seq, longest_orf_rna_length\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8362ebd6",
   "metadata": {},
   "source": [
    "## Isoelectric Point"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bc28b6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cal_protein_features_pi(seq):\n",
    "    '''\n",
    "    :param seq: input primary sequence of a transcript\n",
    "    :return: pi: (protein isoelectric),\n",
    "             mw: (molecular weight)\n",
    "             # gravy: (grand average of hydropathy)\n",
    "             aromaticity: (relative frequency of Phe+Trp+Tyr)\n",
    "             instability: (instability index --\n",
    "                          Any value above 40 means the protein is unstable (=has a short half life)).\n",
    "    '''\n",
    "    protein_seq = translation(seq).strip('*')\n",
    "    protein_object = ProtParam.ProteinAnalysis(protein_seq)\n",
    "    pi = protein_object.isoelectric_point()\n",
    "    return pi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e396d15c",
   "metadata": {},
   "source": [
    "## Aromaticity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f506f788",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cal_protein_features_aromaticity(seq):\n",
    "    '''\n",
    "    :param seq: input primary sequence of a transcript\n",
    "    :return: pi: (protein isoelectric),\n",
    "             mw: (molecular weight)\n",
    "             # gravy: (grand average of hydropathy)\n",
    "             aromaticity: (relative frequency of Phe+Trp+Tyr)\n",
    "             instability: (instability index --\n",
    "                          Any value above 40 means the protein is unstable (=has a short half life)).\n",
    "    '''\n",
    "\n",
    "    protein_seq = translation(seq).strip('*')\n",
    "    protein_object = ProtParam.ProteinAnalysis(protein_seq)\n",
    "    aromaticity = protein_object.aromaticity()\n",
    "    return aromaticity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0096e77e",
   "metadata": {},
   "source": [
    "## Transcript Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c7fc594d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcript_length(seq):\n",
    "    return len(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a81e9003",
   "metadata": {},
   "source": [
    "## CpG Islands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fd8a4065",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_cpg_islands(seq):\n",
    "    cpg_islands = []\n",
    "    cpg_count = 0\n",
    "    in_island = False\n",
    "    \n",
    "    for i in range(len(seq)-1):\n",
    "        if seq[i:i+2].upper() == 'CG':\n",
    "            cpg_count += 1\n",
    "            if not in_island:\n",
    "                in_island = True\n",
    "        else:\n",
    "            if cpg_count > 0:\n",
    "                cpg_islands.append(cpg_count)\n",
    "                cpg_count = 0\n",
    "                in_island = False\n",
    "    \n",
    "    if cpg_count > 0:\n",
    "        cpg_islands.append(cpg_count)\n",
    "    \n",
    "    return cpg_islands"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5365a85f",
   "metadata": {},
   "source": [
    "## Stop Codon Frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f76a43b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stop_codon_num(seq):\n",
    "    translate_prot = Seq(seq).translate()\n",
    "    stop_num = translate_prot.count(\"*\")\n",
    "    return stop_num\n",
    "\n",
    "def get_stop_codon_frequency(seq):\n",
    "    stop_num = get_stop_codon_num(seq)\n",
    "    transript_length = np.log(len(seq)+1)\n",
    "    stop_freq = float(stop_num) / transript_length\n",
    "    return stop_freq\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3afc1de",
   "metadata": {},
   "source": [
    "## Molecular Weight"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b73b0275",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_Mw(seq):\n",
    "    seqprot = Seq(seq).translate()\n",
    "    strinfoAmbiguous = re.compile(\"X|B|Z|J|U\", re.I)\n",
    "    newseqprot = strinfoAmbiguous.sub(\"\", str(seqprot))\n",
    "    protparam_obj = ProtParam.ProteinAnalysis(str(newseqprot).replace(\"*\",\"\"))\n",
    "    mw = protparam_obj.molecular_weight()\n",
    "    return mw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d22cae",
   "metadata": {},
   "source": [
    "## Gravy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c652cfcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gravy(seq):\n",
    "    seqprot = Seq(seq).translate()\n",
    "    strinfoAmbiguous = re.compile(\"X|B|Z|J|U\", re.I)\n",
    "    newseqprot = strinfoAmbiguous.sub(\"\", str(seqprot))\n",
    "    protparam_obj = ProtParam.ProteinAnalysis(str(newseqprot).replace(\"*\", \"\"))\n",
    "    Gravy = protparam_obj.gravy()\n",
    "    return Gravy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d44bd09",
   "metadata": {},
   "source": [
    "## Insatbility Index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bcd5a79e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_instablility_index(seq):\n",
    "    seqprot = Seq(seq).translate()\n",
    "    strinfoAmbiguous = re.compile(\"X|B|Z|J|U\", re.I)\n",
    "    newseqprot = strinfoAmbiguous.sub(\"\", str(seqprot))\n",
    "    protparam_obj = ProtParam.ProteinAnalysis(str(newseqprot).replace(\"*\", \"\"))\n",
    "    instablility_index = protparam_obj.instability_index()\n",
    "    return instablility_index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b90601",
   "metadata": {},
   "source": [
    "## Stop Frame Score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "059f07e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stop_frame_score(seq):\n",
    "    stop_num_in_frame1 = get_stop_codon_num(seq)\n",
    "    stop_num_in_frame2 = get_stop_codon_num(seq[1:])\n",
    "    stop_num_in_frame3 = get_stop_codon_num(seq[2:])\n",
    "    stop_num_all = [stop_num_in_frame1, stop_num_in_frame2, stop_num_in_frame3]\n",
    "    stop_num_frame = ((stop_num_all[0] - stop_num_all[1]) ** 2 + (stop_num_all[0] - stop_num_all[2]) ** 2 + (\n",
    "            stop_num_all[1] - stop_num_all[2]) ** 2) / 2\n",
    "    return stop_num_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "114ae374",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8a16bac",
   "metadata": {},
   "source": [
    "## Generating the Feature Arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "68bb367c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_feature(list_name):\n",
    "    \n",
    "    \n",
    "    for header, sequence in list_name:        \n",
    "        #calculate fickett score\n",
    "        type_seq = 1\n",
    "        fickett_score = fickett_value_full_sequence(sequence, type_seq)\n",
    "        \n",
    "        \n",
    "        #Nucleotide Frequencies\n",
    "        nucleiA = accumulated_frequency_A(sequence)\n",
    "        nucleiG = accumulated_frequency_G(sequence)\n",
    "        nucleiT = accumulated_frequency_T(sequence)\n",
    "        nucleiC = accumulated_frequency_C(sequence)\n",
    "        \n",
    "        #calculate GC content\n",
    "        gc = gcContent(sequence)\n",
    "        \n",
    "        # Calculate the number of CpG islands for this sequence\n",
    "        cpg_islands = calculate_cpg_islands(sequence)\n",
    "        num_islands = len(cpg_islands)\n",
    "    \n",
    "        #calculate transcript length\n",
    "        length = transcript_length(sequence)\n",
    "        \n",
    "        #calculate ORF length\n",
    "        orf_length = ORFfinder_T0(sequence)\n",
    "        \n",
    "        #stop codon freqeuncy\n",
    "        scf = get_stop_codon_frequency(sequence)\n",
    "        \n",
    "        #molecular weight\n",
    "        mw = get_Mw(sequence)\n",
    "        \n",
    "        #gravy\n",
    "        gv = get_gravy(sequence)\n",
    "        \n",
    "        #instability index\n",
    "        ii = get_instablility_index(sequence)\n",
    "        \n",
    "        # stop frame score\n",
    "        sframesc = get_stop_frame_score(sequence)\n",
    "        \n",
    "        #calculate Relative Codon Bias , isoelectric point, aromaticity\n",
    "        if orf_length != 0:\n",
    "            rcb = RCB_score(sequence)\n",
    "            pi = cal_protein_features_pi(sequence)\n",
    "            arom = cal_protein_features_aromaticity(sequence)\n",
    "        else:\n",
    "            rcb = 0\n",
    "            pi = 0\n",
    "            arom = 0\n",
    "        \n",
    "        fickett_array.append(fickett_score)\n",
    "        nucleiA_array.append(nucleiA)\n",
    "        nucleiG_array.append(nucleiG)\n",
    "        nucleiC_array.append(nucleiC)\n",
    "        nucleiT_array.append(nucleiT)\n",
    "        gc_array.append(gc)\n",
    "        cpg_array.append(num_islands)\n",
    "        transcript_array.append(length)\n",
    "        orf_array.append(orf_length)\n",
    "        rcb_array.append(rcb)\n",
    "        ip_array.append(pi)\n",
    "        scf_array.append(scf)\n",
    "        mw_array.append(mw)\n",
    "        gv_array.append(gv)\n",
    "        ii_array.append(ii)\n",
    "        sframesc_array.append(sframesc)\n",
    "        aromaticity_array.append(arom)   \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57db964e",
   "metadata": {},
   "source": [
    "## Generating Coding Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0c49a3f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jae-silver/anaconda3/lib/python3.10/site-packages/Bio/Seq.py:2979: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "fickett_array = []\n",
    "nucleiA_array = []\n",
    "nucleiG_array = []\n",
    "nucleiT_array = []\n",
    "nucleiC_array = []\n",
    "gc_array = []\n",
    "orf_array = []\n",
    "rcb_array=[]\n",
    "ip_array = []\n",
    "aromaticity_array = []\n",
    "transcript_array = []\n",
    "cpg_array = []\n",
    "scf_array = []\n",
    "mw_array = []\n",
    "gv_array = []\n",
    "ii_array = []\n",
    "sframesc_array = []\n",
    "generate_feature(coding_list) # generate feature set for coding dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "223fc1b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_dict = {'Fickett Score':fickett_array, 'NucleiA':nucleiA_array, 'NucleiG':nucleiG_array, 'NucleiT':nucleiT_array, 'NucleiC':nucleiC_array, 'GC content':gc_array, 'ORF Length':orf_array, 'Relative Codon Bias':rcb_array, 'Isoelectric Potential':ip_array, 'Aromaticity':aromaticity_array, 'Transcript Length':transcript_array, 'CpG Islands':cpg_array, 'SCF':scf_array, 'MW': mw_array, 'Gravy': gv_array, 'Instability Index': ii_array, 'Stop Frame': sframesc_array}\n",
    "features_df = pd.DataFrame(names_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "97c985c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fickett Score</th>\n",
       "      <th>NucleiA</th>\n",
       "      <th>NucleiG</th>\n",
       "      <th>NucleiT</th>\n",
       "      <th>NucleiC</th>\n",
       "      <th>GC content</th>\n",
       "      <th>ORF Length</th>\n",
       "      <th>Relative Codon Bias</th>\n",
       "      <th>Isoelectric Potential</th>\n",
       "      <th>Aromaticity</th>\n",
       "      <th>Transcript Length</th>\n",
       "      <th>CpG Islands</th>\n",
       "      <th>SCF</th>\n",
       "      <th>MW</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Instability Index</th>\n",
       "      <th>Stop Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.37617</td>\n",
       "      <td>0.321238</td>\n",
       "      <td>0.175707</td>\n",
       "      <td>0.319328</td>\n",
       "      <td>0.183728</td>\n",
       "      <td>0.359435</td>\n",
       "      <td>981</td>\n",
       "      <td>0.284229</td>\n",
       "      <td>9.588915</td>\n",
       "      <td>0.112385</td>\n",
       "      <td>2618</td>\n",
       "      <td>18</td>\n",
       "      <td>6.225742</td>\n",
       "      <td>95048.7649</td>\n",
       "      <td>0.173633</td>\n",
       "      <td>39.158603</td>\n",
       "      <td>112.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.46720</td>\n",
       "      <td>0.209798</td>\n",
       "      <td>0.202343</td>\n",
       "      <td>0.330138</td>\n",
       "      <td>0.257721</td>\n",
       "      <td>0.460064</td>\n",
       "      <td>939</td>\n",
       "      <td>0.460470</td>\n",
       "      <td>9.042610</td>\n",
       "      <td>0.137821</td>\n",
       "      <td>939</td>\n",
       "      <td>8</td>\n",
       "      <td>0.146073</td>\n",
       "      <td>35073.4005</td>\n",
       "      <td>0.755449</td>\n",
       "      <td>37.706122</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.46720</td>\n",
       "      <td>0.209798</td>\n",
       "      <td>0.202343</td>\n",
       "      <td>0.330138</td>\n",
       "      <td>0.257721</td>\n",
       "      <td>0.460064</td>\n",
       "      <td>939</td>\n",
       "      <td>0.460470</td>\n",
       "      <td>9.042610</td>\n",
       "      <td>0.137821</td>\n",
       "      <td>939</td>\n",
       "      <td>8</td>\n",
       "      <td>0.146073</td>\n",
       "      <td>35073.4005</td>\n",
       "      <td>0.755449</td>\n",
       "      <td>37.706122</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.41608</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>0.341703</td>\n",
       "      <td>0.139105</td>\n",
       "      <td>0.357576</td>\n",
       "      <td>0.699278</td>\n",
       "      <td>2535</td>\n",
       "      <td>0.270346</td>\n",
       "      <td>11.870000</td>\n",
       "      <td>0.051948</td>\n",
       "      <td>3465</td>\n",
       "      <td>302</td>\n",
       "      <td>2.821824</td>\n",
       "      <td>120496.2512</td>\n",
       "      <td>-0.734364</td>\n",
       "      <td>77.579329</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.41608</td>\n",
       "      <td>0.161765</td>\n",
       "      <td>0.341696</td>\n",
       "      <td>0.138985</td>\n",
       "      <td>0.357555</td>\n",
       "      <td>0.699250</td>\n",
       "      <td>2538</td>\n",
       "      <td>0.271500</td>\n",
       "      <td>11.870000</td>\n",
       "      <td>0.051903</td>\n",
       "      <td>3468</td>\n",
       "      <td>302</td>\n",
       "      <td>2.821524</td>\n",
       "      <td>120624.3804</td>\n",
       "      <td>-0.736805</td>\n",
       "      <td>77.440600</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fickett Score   NucleiA   NucleiG   NucleiT   NucleiC  GC content  \\\n",
       "0        0.37617  0.321238  0.175707  0.319328  0.183728    0.359435   \n",
       "1        0.46720  0.209798  0.202343  0.330138  0.257721    0.460064   \n",
       "2        0.46720  0.209798  0.202343  0.330138  0.257721    0.460064   \n",
       "3        0.41608  0.161616  0.341703  0.139105  0.357576    0.699278   \n",
       "4        0.41608  0.161765  0.341696  0.138985  0.357555    0.699250   \n",
       "\n",
       "   ORF Length  Relative Codon Bias  Isoelectric Potential  Aromaticity  \\\n",
       "0         981             0.284229               9.588915     0.112385   \n",
       "1         939             0.460470               9.042610     0.137821   \n",
       "2         939             0.460470               9.042610     0.137821   \n",
       "3        2535             0.270346              11.870000     0.051948   \n",
       "4        2538             0.271500              11.870000     0.051903   \n",
       "\n",
       "   Transcript Length  CpG Islands       SCF           MW     Gravy  \\\n",
       "0               2618           18  6.225742   95048.7649  0.173633   \n",
       "1                939            8  0.146073   35073.4005  0.755449   \n",
       "2                939            8  0.146073   35073.4005  0.755449   \n",
       "3               3465          302  2.821824  120496.2512 -0.734364   \n",
       "4               3468          302  2.821524  120624.3804 -0.736805   \n",
       "\n",
       "   Instability Index  Stop Frame  \n",
       "0          39.158603       112.0  \n",
       "1          37.706122       172.0  \n",
       "2          37.706122       172.0  \n",
       "3          77.579329       252.0  \n",
       "4          77.440600       252.0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f0447ab2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fickett Score</th>\n",
       "      <th>NucleiA</th>\n",
       "      <th>NucleiG</th>\n",
       "      <th>NucleiT</th>\n",
       "      <th>NucleiC</th>\n",
       "      <th>GC content</th>\n",
       "      <th>ORF Length</th>\n",
       "      <th>Relative Codon Bias</th>\n",
       "      <th>Isoelectric Potential</th>\n",
       "      <th>Aromaticity</th>\n",
       "      <th>Transcript Length</th>\n",
       "      <th>CpG Islands</th>\n",
       "      <th>SCF</th>\n",
       "      <th>MW</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Instability Index</th>\n",
       "      <th>Stop Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.37617</td>\n",
       "      <td>0.321238</td>\n",
       "      <td>0.175707</td>\n",
       "      <td>0.319328</td>\n",
       "      <td>0.183728</td>\n",
       "      <td>0.359435</td>\n",
       "      <td>981</td>\n",
       "      <td>0.284229</td>\n",
       "      <td>9.588915</td>\n",
       "      <td>0.112385</td>\n",
       "      <td>2618</td>\n",
       "      <td>18</td>\n",
       "      <td>6.225742</td>\n",
       "      <td>95048.7649</td>\n",
       "      <td>0.173633</td>\n",
       "      <td>39.158603</td>\n",
       "      <td>112.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.46720</td>\n",
       "      <td>0.209798</td>\n",
       "      <td>0.202343</td>\n",
       "      <td>0.330138</td>\n",
       "      <td>0.257721</td>\n",
       "      <td>0.460064</td>\n",
       "      <td>939</td>\n",
       "      <td>0.460470</td>\n",
       "      <td>9.042610</td>\n",
       "      <td>0.137821</td>\n",
       "      <td>939</td>\n",
       "      <td>8</td>\n",
       "      <td>0.146073</td>\n",
       "      <td>35073.4005</td>\n",
       "      <td>0.755449</td>\n",
       "      <td>37.706122</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.46720</td>\n",
       "      <td>0.209798</td>\n",
       "      <td>0.202343</td>\n",
       "      <td>0.330138</td>\n",
       "      <td>0.257721</td>\n",
       "      <td>0.460064</td>\n",
       "      <td>939</td>\n",
       "      <td>0.460470</td>\n",
       "      <td>9.042610</td>\n",
       "      <td>0.137821</td>\n",
       "      <td>939</td>\n",
       "      <td>8</td>\n",
       "      <td>0.146073</td>\n",
       "      <td>35073.4005</td>\n",
       "      <td>0.755449</td>\n",
       "      <td>37.706122</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.41608</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>0.341703</td>\n",
       "      <td>0.139105</td>\n",
       "      <td>0.357576</td>\n",
       "      <td>0.699278</td>\n",
       "      <td>2535</td>\n",
       "      <td>0.270346</td>\n",
       "      <td>11.870000</td>\n",
       "      <td>0.051948</td>\n",
       "      <td>3465</td>\n",
       "      <td>302</td>\n",
       "      <td>2.821824</td>\n",
       "      <td>120496.2512</td>\n",
       "      <td>-0.734364</td>\n",
       "      <td>77.579329</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.41608</td>\n",
       "      <td>0.161765</td>\n",
       "      <td>0.341696</td>\n",
       "      <td>0.138985</td>\n",
       "      <td>0.357555</td>\n",
       "      <td>0.699250</td>\n",
       "      <td>2538</td>\n",
       "      <td>0.271500</td>\n",
       "      <td>11.870000</td>\n",
       "      <td>0.051903</td>\n",
       "      <td>3468</td>\n",
       "      <td>302</td>\n",
       "      <td>2.821524</td>\n",
       "      <td>120624.3804</td>\n",
       "      <td>-0.736805</td>\n",
       "      <td>77.440600</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111270</th>\n",
       "      <td>0.42861</td>\n",
       "      <td>0.294798</td>\n",
       "      <td>0.106936</td>\n",
       "      <td>0.303468</td>\n",
       "      <td>0.294798</td>\n",
       "      <td>0.401734</td>\n",
       "      <td>54</td>\n",
       "      <td>0.647228</td>\n",
       "      <td>4.454040</td>\n",
       "      <td>0.095652</td>\n",
       "      <td>346</td>\n",
       "      <td>9</td>\n",
       "      <td>0.683840</td>\n",
       "      <td>12314.8142</td>\n",
       "      <td>1.224324</td>\n",
       "      <td>49.757658</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111271</th>\n",
       "      <td>0.42696</td>\n",
       "      <td>0.282828</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.286195</td>\n",
       "      <td>0.309764</td>\n",
       "      <td>0.430976</td>\n",
       "      <td>297</td>\n",
       "      <td>0.843020</td>\n",
       "      <td>5.727063</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>297</td>\n",
       "      <td>6</td>\n",
       "      <td>0.175528</td>\n",
       "      <td>10578.7250</td>\n",
       "      <td>1.526531</td>\n",
       "      <td>42.487755</td>\n",
       "      <td>169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111272</th>\n",
       "      <td>0.41737</td>\n",
       "      <td>0.301887</td>\n",
       "      <td>0.099419</td>\n",
       "      <td>0.255443</td>\n",
       "      <td>0.343251</td>\n",
       "      <td>0.442671</td>\n",
       "      <td>300</td>\n",
       "      <td>0.423575</td>\n",
       "      <td>9.402150</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>1378</td>\n",
       "      <td>31</td>\n",
       "      <td>1.659954</td>\n",
       "      <td>48912.9066</td>\n",
       "      <td>0.910515</td>\n",
       "      <td>36.584609</td>\n",
       "      <td>1237.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111273</th>\n",
       "      <td>0.41520</td>\n",
       "      <td>0.304084</td>\n",
       "      <td>0.105960</td>\n",
       "      <td>0.246689</td>\n",
       "      <td>0.343267</td>\n",
       "      <td>0.449227</td>\n",
       "      <td>201</td>\n",
       "      <td>0.358543</td>\n",
       "      <td>9.144276</td>\n",
       "      <td>0.091211</td>\n",
       "      <td>1812</td>\n",
       "      <td>40</td>\n",
       "      <td>1.599416</td>\n",
       "      <td>64562.5371</td>\n",
       "      <td>0.717568</td>\n",
       "      <td>31.892111</td>\n",
       "      <td>1597.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111274</th>\n",
       "      <td>0.40045</td>\n",
       "      <td>0.196190</td>\n",
       "      <td>0.356190</td>\n",
       "      <td>0.377143</td>\n",
       "      <td>0.070476</td>\n",
       "      <td>0.426667</td>\n",
       "      <td>372</td>\n",
       "      <td>0.624221</td>\n",
       "      <td>4.311204</td>\n",
       "      <td>0.120000</td>\n",
       "      <td>525</td>\n",
       "      <td>8</td>\n",
       "      <td>0.478828</td>\n",
       "      <td>18183.3637</td>\n",
       "      <td>1.102907</td>\n",
       "      <td>23.455814</td>\n",
       "      <td>367.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111275 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Fickett Score   NucleiA   NucleiG   NucleiT   NucleiC  GC content  \\\n",
       "0             0.37617  0.321238  0.175707  0.319328  0.183728    0.359435   \n",
       "1             0.46720  0.209798  0.202343  0.330138  0.257721    0.460064   \n",
       "2             0.46720  0.209798  0.202343  0.330138  0.257721    0.460064   \n",
       "3             0.41608  0.161616  0.341703  0.139105  0.357576    0.699278   \n",
       "4             0.41608  0.161765  0.341696  0.138985  0.357555    0.699250   \n",
       "...               ...       ...       ...       ...       ...         ...   \n",
       "111270        0.42861  0.294798  0.106936  0.303468  0.294798    0.401734   \n",
       "111271        0.42696  0.282828  0.121212  0.286195  0.309764    0.430976   \n",
       "111272        0.41737  0.301887  0.099419  0.255443  0.343251    0.442671   \n",
       "111273        0.41520  0.304084  0.105960  0.246689  0.343267    0.449227   \n",
       "111274        0.40045  0.196190  0.356190  0.377143  0.070476    0.426667   \n",
       "\n",
       "        ORF Length  Relative Codon Bias  Isoelectric Potential  Aromaticity  \\\n",
       "0              981             0.284229               9.588915     0.112385   \n",
       "1              939             0.460470               9.042610     0.137821   \n",
       "2              939             0.460470               9.042610     0.137821   \n",
       "3             2535             0.270346              11.870000     0.051948   \n",
       "4             2538             0.271500              11.870000     0.051903   \n",
       "...            ...                  ...                    ...          ...   \n",
       "111270          54             0.647228               4.454040     0.095652   \n",
       "111271         297             0.843020               5.727063     0.071429   \n",
       "111272         300             0.423575               9.402150     0.074074   \n",
       "111273         201             0.358543               9.144276     0.091211   \n",
       "111274         372             0.624221               4.311204     0.120000   \n",
       "\n",
       "        Transcript Length  CpG Islands       SCF           MW     Gravy  \\\n",
       "0                    2618           18  6.225742   95048.7649  0.173633   \n",
       "1                     939            8  0.146073   35073.4005  0.755449   \n",
       "2                     939            8  0.146073   35073.4005  0.755449   \n",
       "3                    3465          302  2.821824  120496.2512 -0.734364   \n",
       "4                    3468          302  2.821524  120624.3804 -0.736805   \n",
       "...                   ...          ...       ...          ...       ...   \n",
       "111270                346            9  0.683840   12314.8142  1.224324   \n",
       "111271                297            6  0.175528   10578.7250  1.526531   \n",
       "111272               1378           31  1.659954   48912.9066  0.910515   \n",
       "111273               1812           40  1.599416   64562.5371  0.717568   \n",
       "111274                525            8  0.478828   18183.3637  1.102907   \n",
       "\n",
       "        Instability Index  Stop Frame  \n",
       "0               39.158603       112.0  \n",
       "1               37.706122       172.0  \n",
       "2               37.706122       172.0  \n",
       "3               77.579329       252.0  \n",
       "4               77.440600       252.0  \n",
       "...                   ...         ...  \n",
       "111270          49.757658       211.0  \n",
       "111271          42.487755       169.0  \n",
       "111272          36.584609      1237.0  \n",
       "111273          31.892111      1597.0  \n",
       "111274          23.455814       367.0  \n",
       "\n",
       "[111275 rows x 17 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e833857",
   "metadata": {},
   "source": [
    "### Adding a label for coding sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3f5212f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fickett Score</th>\n",
       "      <th>NucleiA</th>\n",
       "      <th>NucleiG</th>\n",
       "      <th>NucleiT</th>\n",
       "      <th>NucleiC</th>\n",
       "      <th>GC content</th>\n",
       "      <th>ORF Length</th>\n",
       "      <th>Relative Codon Bias</th>\n",
       "      <th>Isoelectric Potential</th>\n",
       "      <th>Aromaticity</th>\n",
       "      <th>Transcript Length</th>\n",
       "      <th>CpG Islands</th>\n",
       "      <th>SCF</th>\n",
       "      <th>MW</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Instability Index</th>\n",
       "      <th>Stop Frame</th>\n",
       "      <th>coding/noncoding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.37617</td>\n",
       "      <td>0.321238</td>\n",
       "      <td>0.175707</td>\n",
       "      <td>0.319328</td>\n",
       "      <td>0.183728</td>\n",
       "      <td>0.359435</td>\n",
       "      <td>981</td>\n",
       "      <td>0.284229</td>\n",
       "      <td>9.588915</td>\n",
       "      <td>0.112385</td>\n",
       "      <td>2618</td>\n",
       "      <td>18</td>\n",
       "      <td>6.225742</td>\n",
       "      <td>95048.7649</td>\n",
       "      <td>0.173633</td>\n",
       "      <td>39.158603</td>\n",
       "      <td>112.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.46720</td>\n",
       "      <td>0.209798</td>\n",
       "      <td>0.202343</td>\n",
       "      <td>0.330138</td>\n",
       "      <td>0.257721</td>\n",
       "      <td>0.460064</td>\n",
       "      <td>939</td>\n",
       "      <td>0.460470</td>\n",
       "      <td>9.042610</td>\n",
       "      <td>0.137821</td>\n",
       "      <td>939</td>\n",
       "      <td>8</td>\n",
       "      <td>0.146073</td>\n",
       "      <td>35073.4005</td>\n",
       "      <td>0.755449</td>\n",
       "      <td>37.706122</td>\n",
       "      <td>172.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.46720</td>\n",
       "      <td>0.209798</td>\n",
       "      <td>0.202343</td>\n",
       "      <td>0.330138</td>\n",
       "      <td>0.257721</td>\n",
       "      <td>0.460064</td>\n",
       "      <td>939</td>\n",
       "      <td>0.460470</td>\n",
       "      <td>9.042610</td>\n",
       "      <td>0.137821</td>\n",
       "      <td>939</td>\n",
       "      <td>8</td>\n",
       "      <td>0.146073</td>\n",
       "      <td>35073.4005</td>\n",
       "      <td>0.755449</td>\n",
       "      <td>37.706122</td>\n",
       "      <td>172.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.41608</td>\n",
       "      <td>0.161616</td>\n",
       "      <td>0.341703</td>\n",
       "      <td>0.139105</td>\n",
       "      <td>0.357576</td>\n",
       "      <td>0.699278</td>\n",
       "      <td>2535</td>\n",
       "      <td>0.270346</td>\n",
       "      <td>11.870000</td>\n",
       "      <td>0.051948</td>\n",
       "      <td>3465</td>\n",
       "      <td>302</td>\n",
       "      <td>2.821824</td>\n",
       "      <td>120496.2512</td>\n",
       "      <td>-0.734364</td>\n",
       "      <td>77.579329</td>\n",
       "      <td>252.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.41608</td>\n",
       "      <td>0.161765</td>\n",
       "      <td>0.341696</td>\n",
       "      <td>0.138985</td>\n",
       "      <td>0.357555</td>\n",
       "      <td>0.699250</td>\n",
       "      <td>2538</td>\n",
       "      <td>0.271500</td>\n",
       "      <td>11.870000</td>\n",
       "      <td>0.051903</td>\n",
       "      <td>3468</td>\n",
       "      <td>302</td>\n",
       "      <td>2.821524</td>\n",
       "      <td>120624.3804</td>\n",
       "      <td>-0.736805</td>\n",
       "      <td>77.440600</td>\n",
       "      <td>252.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fickett Score   NucleiA   NucleiG   NucleiT   NucleiC  GC content  \\\n",
       "0        0.37617  0.321238  0.175707  0.319328  0.183728    0.359435   \n",
       "1        0.46720  0.209798  0.202343  0.330138  0.257721    0.460064   \n",
       "2        0.46720  0.209798  0.202343  0.330138  0.257721    0.460064   \n",
       "3        0.41608  0.161616  0.341703  0.139105  0.357576    0.699278   \n",
       "4        0.41608  0.161765  0.341696  0.138985  0.357555    0.699250   \n",
       "\n",
       "   ORF Length  Relative Codon Bias  Isoelectric Potential  Aromaticity  \\\n",
       "0         981             0.284229               9.588915     0.112385   \n",
       "1         939             0.460470               9.042610     0.137821   \n",
       "2         939             0.460470               9.042610     0.137821   \n",
       "3        2535             0.270346              11.870000     0.051948   \n",
       "4        2538             0.271500              11.870000     0.051903   \n",
       "\n",
       "   Transcript Length  CpG Islands       SCF           MW     Gravy  \\\n",
       "0               2618           18  6.225742   95048.7649  0.173633   \n",
       "1                939            8  0.146073   35073.4005  0.755449   \n",
       "2                939            8  0.146073   35073.4005  0.755449   \n",
       "3               3465          302  2.821824  120496.2512 -0.734364   \n",
       "4               3468          302  2.821524  120624.3804 -0.736805   \n",
       "\n",
       "   Instability Index  Stop Frame  coding/noncoding  \n",
       "0          39.158603       112.0                 1  \n",
       "1          37.706122       172.0                 1  \n",
       "2          37.706122       172.0                 1  \n",
       "3          77.579329       252.0                 1  \n",
       "4          77.440600       252.0                 1  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df[\"coding/noncoding\"] = 1    #1 for coding sequences\n",
    "features_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "153c6f84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(111275, 18)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d26e37",
   "metadata": {},
   "source": [
    "### To csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7db7e739",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_df.to_csv('coding_17_features.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61d9761d",
   "metadata": {},
   "source": [
    "## Generating Non-Coding Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "62b5ff60",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jae-silver/anaconda3/lib/python3.10/site-packages/Bio/Seq.py:2979: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "fickett_array = []\n",
    "nucleiA_array = []\n",
    "nucleiG_array = []\n",
    "nucleiT_array = []\n",
    "nucleiC_array = []\n",
    "gc_array = []\n",
    "orf_array = []\n",
    "rcb_array=[]\n",
    "ip_array = []\n",
    "aromaticity_array = []\n",
    "transcript_array = []\n",
    "cpg_array = []\n",
    "scf_array = []\n",
    "mw_array = []\n",
    "gv_array = []\n",
    "ii_array = []\n",
    "sframesc_array = []\n",
    "generate_feature(noncoding_list) # generate feature set for coding dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "46bd1848",
   "metadata": {},
   "outputs": [],
   "source": [
    "names_dict = {'Fickett Score':fickett_array, 'NucleiA':nucleiA_array, 'NucleiG':nucleiG_array, 'NucleiT':nucleiT_array, 'NucleiC':nucleiC_array, 'GC content':gc_array, 'ORF Length':orf_array, 'Relative Codon Bias':rcb_array, 'Isoelectric Potential':ip_array, 'Aromaticity':aromaticity_array, 'Transcript Length':transcript_array, 'CpG Islands':cpg_array, 'SCF':scf_array, 'MW': mw_array, 'Gravy': gv_array, 'Instability Index': ii_array, 'Stop Frame': sframesc_array}\n",
    "features_df = pd.DataFrame(names_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9df3f1df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fickett Score</th>\n",
       "      <th>NucleiA</th>\n",
       "      <th>NucleiG</th>\n",
       "      <th>NucleiT</th>\n",
       "      <th>NucleiC</th>\n",
       "      <th>GC content</th>\n",
       "      <th>ORF Length</th>\n",
       "      <th>Relative Codon Bias</th>\n",
       "      <th>Isoelectric Potential</th>\n",
       "      <th>Aromaticity</th>\n",
       "      <th>Transcript Length</th>\n",
       "      <th>CpG Islands</th>\n",
       "      <th>SCF</th>\n",
       "      <th>MW</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Instability Index</th>\n",
       "      <th>Stop Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.33022</td>\n",
       "      <td>0.192982</td>\n",
       "      <td>0.245009</td>\n",
       "      <td>0.253479</td>\n",
       "      <td>0.308530</td>\n",
       "      <td>0.553539</td>\n",
       "      <td>390</td>\n",
       "      <td>0.466010</td>\n",
       "      <td>8.557163</td>\n",
       "      <td>0.079855</td>\n",
       "      <td>1653</td>\n",
       "      <td>23</td>\n",
       "      <td>1.754161</td>\n",
       "      <td>58803.5325</td>\n",
       "      <td>0.013569</td>\n",
       "      <td>69.028104</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.33326</td>\n",
       "      <td>0.193527</td>\n",
       "      <td>0.241403</td>\n",
       "      <td>0.253540</td>\n",
       "      <td>0.311531</td>\n",
       "      <td>0.552933</td>\n",
       "      <td>402</td>\n",
       "      <td>0.428935</td>\n",
       "      <td>8.525573</td>\n",
       "      <td>0.056680</td>\n",
       "      <td>1483</td>\n",
       "      <td>21</td>\n",
       "      <td>2.464910</td>\n",
       "      <td>50958.2792</td>\n",
       "      <td>-0.037395</td>\n",
       "      <td>61.601050</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.34850</td>\n",
       "      <td>0.213608</td>\n",
       "      <td>0.299051</td>\n",
       "      <td>0.205696</td>\n",
       "      <td>0.281646</td>\n",
       "      <td>0.580696</td>\n",
       "      <td>180</td>\n",
       "      <td>0.591729</td>\n",
       "      <td>8.961508</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>632</td>\n",
       "      <td>10</td>\n",
       "      <td>0.930165</td>\n",
       "      <td>21749.5998</td>\n",
       "      <td>-0.254902</td>\n",
       "      <td>72.012255</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.34765</td>\n",
       "      <td>0.218935</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.163708</td>\n",
       "      <td>0.284024</td>\n",
       "      <td>0.617357</td>\n",
       "      <td>165</td>\n",
       "      <td>0.659118</td>\n",
       "      <td>11.371079</td>\n",
       "      <td>0.059172</td>\n",
       "      <td>507</td>\n",
       "      <td>18</td>\n",
       "      <td>0.321002</td>\n",
       "      <td>18519.8711</td>\n",
       "      <td>-0.794012</td>\n",
       "      <td>70.119162</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.34143</td>\n",
       "      <td>0.200330</td>\n",
       "      <td>0.281121</td>\n",
       "      <td>0.201979</td>\n",
       "      <td>0.316570</td>\n",
       "      <td>0.597692</td>\n",
       "      <td>165</td>\n",
       "      <td>0.529159</td>\n",
       "      <td>10.728845</td>\n",
       "      <td>0.059406</td>\n",
       "      <td>1213</td>\n",
       "      <td>34</td>\n",
       "      <td>1.689742</td>\n",
       "      <td>41313.5081</td>\n",
       "      <td>-0.344133</td>\n",
       "      <td>75.329872</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fickett Score   NucleiA   NucleiG   NucleiT   NucleiC  GC content  \\\n",
       "0        0.33022  0.192982  0.245009  0.253479  0.308530    0.553539   \n",
       "1        0.33326  0.193527  0.241403  0.253540  0.311531    0.552933   \n",
       "2        0.34850  0.213608  0.299051  0.205696  0.281646    0.580696   \n",
       "3        0.34765  0.218935  0.333333  0.163708  0.284024    0.617357   \n",
       "4        0.34143  0.200330  0.281121  0.201979  0.316570    0.597692   \n",
       "\n",
       "   ORF Length  Relative Codon Bias  Isoelectric Potential  Aromaticity  \\\n",
       "0         390             0.466010               8.557163     0.079855   \n",
       "1         402             0.428935               8.525573     0.056680   \n",
       "2         180             0.591729               8.961508     0.057143   \n",
       "3         165             0.659118              11.371079     0.059172   \n",
       "4         165             0.529159              10.728845     0.059406   \n",
       "\n",
       "   Transcript Length  CpG Islands       SCF          MW     Gravy  \\\n",
       "0               1653           23  1.754161  58803.5325  0.013569   \n",
       "1               1483           21  2.464910  50958.2792 -0.037395   \n",
       "2                632           10  0.930165  21749.5998 -0.254902   \n",
       "3                507           18  0.321002  18519.8711 -0.794012   \n",
       "4               1213           34  1.689742  41313.5081 -0.344133   \n",
       "\n",
       "   Instability Index  Stop Frame  \n",
       "0          69.028104         7.0  \n",
       "1          61.601050        57.0  \n",
       "2          72.012255         3.0  \n",
       "3          70.119162         9.0  \n",
       "4          75.329872         3.0  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "12c74b3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fickett Score</th>\n",
       "      <th>NucleiA</th>\n",
       "      <th>NucleiG</th>\n",
       "      <th>NucleiT</th>\n",
       "      <th>NucleiC</th>\n",
       "      <th>GC content</th>\n",
       "      <th>ORF Length</th>\n",
       "      <th>Relative Codon Bias</th>\n",
       "      <th>Isoelectric Potential</th>\n",
       "      <th>Aromaticity</th>\n",
       "      <th>Transcript Length</th>\n",
       "      <th>CpG Islands</th>\n",
       "      <th>SCF</th>\n",
       "      <th>MW</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Instability Index</th>\n",
       "      <th>Stop Frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.33022</td>\n",
       "      <td>0.192982</td>\n",
       "      <td>0.245009</td>\n",
       "      <td>0.253479</td>\n",
       "      <td>0.308530</td>\n",
       "      <td>0.553539</td>\n",
       "      <td>390</td>\n",
       "      <td>0.466010</td>\n",
       "      <td>8.557163</td>\n",
       "      <td>0.079855</td>\n",
       "      <td>1653</td>\n",
       "      <td>23</td>\n",
       "      <td>1.754161</td>\n",
       "      <td>58803.5325</td>\n",
       "      <td>0.013569</td>\n",
       "      <td>69.028104</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.33326</td>\n",
       "      <td>0.193527</td>\n",
       "      <td>0.241403</td>\n",
       "      <td>0.253540</td>\n",
       "      <td>0.311531</td>\n",
       "      <td>0.552933</td>\n",
       "      <td>402</td>\n",
       "      <td>0.428935</td>\n",
       "      <td>8.525573</td>\n",
       "      <td>0.056680</td>\n",
       "      <td>1483</td>\n",
       "      <td>21</td>\n",
       "      <td>2.464910</td>\n",
       "      <td>50958.2792</td>\n",
       "      <td>-0.037395</td>\n",
       "      <td>61.601050</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.34850</td>\n",
       "      <td>0.213608</td>\n",
       "      <td>0.299051</td>\n",
       "      <td>0.205696</td>\n",
       "      <td>0.281646</td>\n",
       "      <td>0.580696</td>\n",
       "      <td>180</td>\n",
       "      <td>0.591729</td>\n",
       "      <td>8.961508</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>632</td>\n",
       "      <td>10</td>\n",
       "      <td>0.930165</td>\n",
       "      <td>21749.5998</td>\n",
       "      <td>-0.254902</td>\n",
       "      <td>72.012255</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.34765</td>\n",
       "      <td>0.218935</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.163708</td>\n",
       "      <td>0.284024</td>\n",
       "      <td>0.617357</td>\n",
       "      <td>165</td>\n",
       "      <td>0.659118</td>\n",
       "      <td>11.371079</td>\n",
       "      <td>0.059172</td>\n",
       "      <td>507</td>\n",
       "      <td>18</td>\n",
       "      <td>0.321002</td>\n",
       "      <td>18519.8711</td>\n",
       "      <td>-0.794012</td>\n",
       "      <td>70.119162</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.34143</td>\n",
       "      <td>0.200330</td>\n",
       "      <td>0.281121</td>\n",
       "      <td>0.201979</td>\n",
       "      <td>0.316570</td>\n",
       "      <td>0.597692</td>\n",
       "      <td>165</td>\n",
       "      <td>0.529159</td>\n",
       "      <td>10.728845</td>\n",
       "      <td>0.059406</td>\n",
       "      <td>1213</td>\n",
       "      <td>34</td>\n",
       "      <td>1.689742</td>\n",
       "      <td>41313.5081</td>\n",
       "      <td>-0.344133</td>\n",
       "      <td>75.329872</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173106</th>\n",
       "      <td>0.34933</td>\n",
       "      <td>0.279215</td>\n",
       "      <td>0.243533</td>\n",
       "      <td>0.214095</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.506690</td>\n",
       "      <td>165</td>\n",
       "      <td>0.404011</td>\n",
       "      <td>9.778000</td>\n",
       "      <td>0.050938</td>\n",
       "      <td>1121</td>\n",
       "      <td>29</td>\n",
       "      <td>1.993488</td>\n",
       "      <td>39898.6091</td>\n",
       "      <td>-0.465181</td>\n",
       "      <td>56.477437</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173107</th>\n",
       "      <td>0.30073</td>\n",
       "      <td>0.267338</td>\n",
       "      <td>0.275168</td>\n",
       "      <td>0.197987</td>\n",
       "      <td>0.259508</td>\n",
       "      <td>0.534676</td>\n",
       "      <td>93</td>\n",
       "      <td>0.439534</td>\n",
       "      <td>9.988038</td>\n",
       "      <td>0.077441</td>\n",
       "      <td>894</td>\n",
       "      <td>38</td>\n",
       "      <td>2.059786</td>\n",
       "      <td>31774.3037</td>\n",
       "      <td>-0.603873</td>\n",
       "      <td>48.504577</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173108</th>\n",
       "      <td>0.31353</td>\n",
       "      <td>0.262545</td>\n",
       "      <td>0.210125</td>\n",
       "      <td>0.218190</td>\n",
       "      <td>0.309140</td>\n",
       "      <td>0.519265</td>\n",
       "      <td>405</td>\n",
       "      <td>0.308874</td>\n",
       "      <td>9.880441</td>\n",
       "      <td>0.084677</td>\n",
       "      <td>2232</td>\n",
       "      <td>89</td>\n",
       "      <td>2.463980</td>\n",
       "      <td>80774.0628</td>\n",
       "      <td>-0.507172</td>\n",
       "      <td>53.483448</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173109</th>\n",
       "      <td>0.30200</td>\n",
       "      <td>0.300735</td>\n",
       "      <td>0.219853</td>\n",
       "      <td>0.269853</td>\n",
       "      <td>0.209559</td>\n",
       "      <td>0.429412</td>\n",
       "      <td>207</td>\n",
       "      <td>0.469681</td>\n",
       "      <td>8.829928</td>\n",
       "      <td>0.092715</td>\n",
       "      <td>1360</td>\n",
       "      <td>8</td>\n",
       "      <td>2.771628</td>\n",
       "      <td>49559.5029</td>\n",
       "      <td>-0.301848</td>\n",
       "      <td>56.339076</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173110</th>\n",
       "      <td>0.32515</td>\n",
       "      <td>0.257373</td>\n",
       "      <td>0.259383</td>\n",
       "      <td>0.230563</td>\n",
       "      <td>0.252681</td>\n",
       "      <td>0.512064</td>\n",
       "      <td>159</td>\n",
       "      <td>0.387534</td>\n",
       "      <td>9.994034</td>\n",
       "      <td>0.078471</td>\n",
       "      <td>1492</td>\n",
       "      <td>55</td>\n",
       "      <td>2.052393</td>\n",
       "      <td>53533.9964</td>\n",
       "      <td>-0.390249</td>\n",
       "      <td>50.519087</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173111 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Fickett Score   NucleiA   NucleiG   NucleiT   NucleiC  GC content  \\\n",
       "0             0.33022  0.192982  0.245009  0.253479  0.308530    0.553539   \n",
       "1             0.33326  0.193527  0.241403  0.253540  0.311531    0.552933   \n",
       "2             0.34850  0.213608  0.299051  0.205696  0.281646    0.580696   \n",
       "3             0.34765  0.218935  0.333333  0.163708  0.284024    0.617357   \n",
       "4             0.34143  0.200330  0.281121  0.201979  0.316570    0.597692   \n",
       "...               ...       ...       ...       ...       ...         ...   \n",
       "173106        0.34933  0.279215  0.243533  0.214095  0.263158    0.506690   \n",
       "173107        0.30073  0.267338  0.275168  0.197987  0.259508    0.534676   \n",
       "173108        0.31353  0.262545  0.210125  0.218190  0.309140    0.519265   \n",
       "173109        0.30200  0.300735  0.219853  0.269853  0.209559    0.429412   \n",
       "173110        0.32515  0.257373  0.259383  0.230563  0.252681    0.512064   \n",
       "\n",
       "        ORF Length  Relative Codon Bias  Isoelectric Potential  Aromaticity  \\\n",
       "0              390             0.466010               8.557163     0.079855   \n",
       "1              402             0.428935               8.525573     0.056680   \n",
       "2              180             0.591729               8.961508     0.057143   \n",
       "3              165             0.659118              11.371079     0.059172   \n",
       "4              165             0.529159              10.728845     0.059406   \n",
       "...            ...                  ...                    ...          ...   \n",
       "173106         165             0.404011               9.778000     0.050938   \n",
       "173107          93             0.439534               9.988038     0.077441   \n",
       "173108         405             0.308874               9.880441     0.084677   \n",
       "173109         207             0.469681               8.829928     0.092715   \n",
       "173110         159             0.387534               9.994034     0.078471   \n",
       "\n",
       "        Transcript Length  CpG Islands       SCF          MW     Gravy  \\\n",
       "0                    1653           23  1.754161  58803.5325  0.013569   \n",
       "1                    1483           21  2.464910  50958.2792 -0.037395   \n",
       "2                     632           10  0.930165  21749.5998 -0.254902   \n",
       "3                     507           18  0.321002  18519.8711 -0.794012   \n",
       "4                    1213           34  1.689742  41313.5081 -0.344133   \n",
       "...                   ...          ...       ...         ...       ...   \n",
       "173106               1121           29  1.993488  39898.6091 -0.465181   \n",
       "173107                894           38  2.059786  31774.3037 -0.603873   \n",
       "173108               2232           89  2.463980  80774.0628 -0.507172   \n",
       "173109               1360            8  2.771628  49559.5029 -0.301848   \n",
       "173110               1492           55  2.052393  53533.9964 -0.390249   \n",
       "\n",
       "        Instability Index  Stop Frame  \n",
       "0               69.028104         7.0  \n",
       "1               61.601050        57.0  \n",
       "2               72.012255         3.0  \n",
       "3               70.119162         9.0  \n",
       "4               75.329872         3.0  \n",
       "...                   ...         ...  \n",
       "173106          56.477437         7.0  \n",
       "173107          48.504577         7.0  \n",
       "173108          53.483448        37.0  \n",
       "173109          56.339076        31.0  \n",
       "173110          50.519087        19.0  \n",
       "\n",
       "[173111 rows x 17 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ffc39ca",
   "metadata": {},
   "source": [
    "### Adding a label for non coding sequences "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5dd3a833",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fickett Score</th>\n",
       "      <th>NucleiA</th>\n",
       "      <th>NucleiG</th>\n",
       "      <th>NucleiT</th>\n",
       "      <th>NucleiC</th>\n",
       "      <th>GC content</th>\n",
       "      <th>ORF Length</th>\n",
       "      <th>Relative Codon Bias</th>\n",
       "      <th>Isoelectric Potential</th>\n",
       "      <th>Aromaticity</th>\n",
       "      <th>Transcript Length</th>\n",
       "      <th>CpG Islands</th>\n",
       "      <th>SCF</th>\n",
       "      <th>MW</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Instability Index</th>\n",
       "      <th>Stop Frame</th>\n",
       "      <th>coding/noncoding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.33022</td>\n",
       "      <td>0.192982</td>\n",
       "      <td>0.245009</td>\n",
       "      <td>0.253479</td>\n",
       "      <td>0.308530</td>\n",
       "      <td>0.553539</td>\n",
       "      <td>390</td>\n",
       "      <td>0.466010</td>\n",
       "      <td>8.557163</td>\n",
       "      <td>0.079855</td>\n",
       "      <td>1653</td>\n",
       "      <td>23</td>\n",
       "      <td>1.754161</td>\n",
       "      <td>58803.5325</td>\n",
       "      <td>0.013569</td>\n",
       "      <td>69.028104</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.33326</td>\n",
       "      <td>0.193527</td>\n",
       "      <td>0.241403</td>\n",
       "      <td>0.253540</td>\n",
       "      <td>0.311531</td>\n",
       "      <td>0.552933</td>\n",
       "      <td>402</td>\n",
       "      <td>0.428935</td>\n",
       "      <td>8.525573</td>\n",
       "      <td>0.056680</td>\n",
       "      <td>1483</td>\n",
       "      <td>21</td>\n",
       "      <td>2.464910</td>\n",
       "      <td>50958.2792</td>\n",
       "      <td>-0.037395</td>\n",
       "      <td>61.601050</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.34850</td>\n",
       "      <td>0.213608</td>\n",
       "      <td>0.299051</td>\n",
       "      <td>0.205696</td>\n",
       "      <td>0.281646</td>\n",
       "      <td>0.580696</td>\n",
       "      <td>180</td>\n",
       "      <td>0.591729</td>\n",
       "      <td>8.961508</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>632</td>\n",
       "      <td>10</td>\n",
       "      <td>0.930165</td>\n",
       "      <td>21749.5998</td>\n",
       "      <td>-0.254902</td>\n",
       "      <td>72.012255</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.34765</td>\n",
       "      <td>0.218935</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.163708</td>\n",
       "      <td>0.284024</td>\n",
       "      <td>0.617357</td>\n",
       "      <td>165</td>\n",
       "      <td>0.659118</td>\n",
       "      <td>11.371079</td>\n",
       "      <td>0.059172</td>\n",
       "      <td>507</td>\n",
       "      <td>18</td>\n",
       "      <td>0.321002</td>\n",
       "      <td>18519.8711</td>\n",
       "      <td>-0.794012</td>\n",
       "      <td>70.119162</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.34143</td>\n",
       "      <td>0.200330</td>\n",
       "      <td>0.281121</td>\n",
       "      <td>0.201979</td>\n",
       "      <td>0.316570</td>\n",
       "      <td>0.597692</td>\n",
       "      <td>165</td>\n",
       "      <td>0.529159</td>\n",
       "      <td>10.728845</td>\n",
       "      <td>0.059406</td>\n",
       "      <td>1213</td>\n",
       "      <td>34</td>\n",
       "      <td>1.689742</td>\n",
       "      <td>41313.5081</td>\n",
       "      <td>-0.344133</td>\n",
       "      <td>75.329872</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Fickett Score   NucleiA   NucleiG   NucleiT   NucleiC  GC content  \\\n",
       "0        0.33022  0.192982  0.245009  0.253479  0.308530    0.553539   \n",
       "1        0.33326  0.193527  0.241403  0.253540  0.311531    0.552933   \n",
       "2        0.34850  0.213608  0.299051  0.205696  0.281646    0.580696   \n",
       "3        0.34765  0.218935  0.333333  0.163708  0.284024    0.617357   \n",
       "4        0.34143  0.200330  0.281121  0.201979  0.316570    0.597692   \n",
       "\n",
       "   ORF Length  Relative Codon Bias  Isoelectric Potential  Aromaticity  \\\n",
       "0         390             0.466010               8.557163     0.079855   \n",
       "1         402             0.428935               8.525573     0.056680   \n",
       "2         180             0.591729               8.961508     0.057143   \n",
       "3         165             0.659118              11.371079     0.059172   \n",
       "4         165             0.529159              10.728845     0.059406   \n",
       "\n",
       "   Transcript Length  CpG Islands       SCF          MW     Gravy  \\\n",
       "0               1653           23  1.754161  58803.5325  0.013569   \n",
       "1               1483           21  2.464910  50958.2792 -0.037395   \n",
       "2                632           10  0.930165  21749.5998 -0.254902   \n",
       "3                507           18  0.321002  18519.8711 -0.794012   \n",
       "4               1213           34  1.689742  41313.5081 -0.344133   \n",
       "\n",
       "   Instability Index  Stop Frame  coding/noncoding  \n",
       "0          69.028104         7.0                 0  \n",
       "1          61.601050        57.0                 0  \n",
       "2          72.012255         3.0                 0  \n",
       "3          70.119162         9.0                 0  \n",
       "4          75.329872         3.0                 0  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df[\"coding/noncoding\"] = 0    #1 for non-coding sequences\n",
    "features_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a96c76b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(173111, 18)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d3491d4",
   "metadata": {},
   "source": [
    "### To CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "92a1d672",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_df.to_csv('noncoding_17_features.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
